### PR TITLE
Raise PHPStan's level from 2 to 3

### PR DESCRIPTION
**Pull request by dkarlovi**
See #513.

### CODE REVIEW

File: Document/AuthCodeManager.php [hunk 1]
```diff
@@ -23,31 +24,31 @@ class AuthCodeManager extends BaseAuthCodeManager
     /**
      * @var DocumentManager
      */
-    protected $dm;
+    protected $documentManager;
 
     /**
-     * @var DocumentRepository
+     * @var ObjectManager|DocumentRepository
```
Comment (dinamic) [inline Document/AuthCodeManager.php]: This seems false.

File: Document/ClientManager.php [hunk 2]
```diff
@@ -23,31 +24,31 @@ class ClientManager extends BaseClientManager
     /**
      * @var DocumentManager
      */
-    protected $dm;
+    protected $documentManager;
 
     /**
-     * @var DocumentRepository
+     * @var ObjectManager|DocumentRepository
```
Comment (dinamic) [inline Document/ClientManager.php]: This one needs fixing IMO. Should be `DocumentRepository` only.

File: Document/TokenManager.php [hunk 3]
```diff
@@ -23,31 +24,31 @@ class TokenManager extends BaseTokenManager
     /**
      * @var DocumentManager
      */
-    protected $dm;
+    protected $documentManager;
 
     /**
-     * @var DocumentRepository
+     * @var ObjectManager|DocumentRepository
```
Comment (dinamic) [inline Document/TokenManager.php]: This seem recurring. Should be `DocumentRepository` IMO.

File: Entity/ClientManager.php [hunk 4]
```diff
@@ -24,31 +24,31 @@ class ClientManager extends BaseClientManager
     /**
      * @var EntityManager
      */
-    protected $em;
+    protected $entityManager;
 
     /**
-     * @var EntityRepository
+     * @var ObjectManager|EntityRepository
```
Comment (dinamic) [inline Entity/ClientManager.php]: This should be set to `EntityRepository` only.

File: Entity/TokenManager.php [hunk 5]
```diff
@@ -24,31 +24,31 @@ class TokenManager extends BaseTokenManager
     /**
      * @var EntityManager
      */
-    protected $em;
+    protected $entityManager;
 
     /**
-     * @var EntityRepository
+     * @var ObjectManager|EntityRepository
```
Comment (dinamic) [inline Entity/TokenManager.php]: Again, `EntityRepository` only.

File: Entity/ClientManager.php (hunk 4 above)
Comment (dkarlovi) [reply Entity/ClientManager.php]: Problem is, `EntityManager::getRepository` is [documented as returning both](https://github.com/doctrine/doctrine2/blob/master/lib/Doctrine/ORM/EntityManager.php#L705).


Asked @ocramius on Twitter, [he thought it's different](https://twitter.com/Ocramius/status/951845101998833664). Changed at https://github.com/doctrine/doctrine2/commit/99db207a9f1214ad483efd81341279a550526358, don't know why.

Comment (Ocramius) [reply Entity/ClientManager.php]: This should be `EntityRepository`. `EntityRepository` implements `ObjectRepository`, so there is no need to change anything here, as the type is compatible.

Comment (dkarlovi) [reply Entity/ClientManager.php]: I think the issue is, you're not guaranteed to get an `EntityRepository` from `EntityManager::getRepository` but an `ObjectRepository`.

That's the reason PHPStan is complaining, might be wrong though. @ondrejmirtes, mind confirming/denying this?

Comment (dinamic) [reply Entity/ClientManager.php]: Still, `ObjectRepository` is not `ObjectManager`. :)

Comment (dkarlovi) [reply Entity/ClientManager.php]: :see_no_evil: now I see what you mean. :+1: That's clearly an error, how on earth did it pass the build.

It's supposed to be `ObjectRepository`.

Comment (JanTvrdik) [reply Entity/ClientManager.php]: Note that `ObjectRepository|EntityRepository` is the same as just saying `ObjectRepository`.

Comment (ondrejmirtes) [reply Entity/ClientManager.php]: @dkarlovi What exactly is PHPStan complaining about?

Comment (dkarlovi) [reply Entity/ClientManager.php]: ObjectRepository cannot be assigned to EntityRepository.

Comment (ondrejmirtes) [reply Entity/ClientManager.php]: Yes, makes sense, so `ObjectRepository` should be used everywhere.

**Comment by dkarlovi** [comment]
@dinamic one test is failing (might be related to PHPUnit's update to v6 so we can use PHPStan's plugin):

```
There was 1 failure:

1) FOS\OAuthServerBundle\Tests\Form\Handler\AuthorizeFormHandlerTest::testProcessWillHandleRequestOnPostAndWillProcessDataIfFormIsValid
Failed asserting that Array &0 (
    'client_id' => Binary String: 0x314a761dc55461a75569
    'response_type' => '�V�R�J�ߐ'
    'redirect_uri' => 'U@�Z����5Y'
    'state' => '�qz����&BW'
    'scope' => Binary String: 0x6a00d193d5427e821133
) is identical to Array &0 ().
```

If you could help me out here, that would be great-

**Comment by dkarlovi** [comment]
I'll just mark it as skipped until we figure out why it's failing, don't want a single test block us from moving forward here as I've incorporated PHPStan's PHPUnit extension.

**Comment by dinamic** [comment]
Let me have a look.

**Comment by dinamic** [comment]
Phpunit v6 does not seem to clean the global scope after each test, that's why the test is failing.

You can workaround this by putting `$_GET = [];` in `setUp()`.

Here's what the official site says about this:

```
Prior to version 6, by default, PHPUnit ran your tests in a way where changes to global and super-global variables ($GLOBALS, $_ENV, $_POST, $_GET, $_COOKIE, $_SERVER, $_FILES, $_REQUEST) do not affect other tests.

As of version 6, PHPUnit does not perform this backup and restore operation for global and super-global variables by default anymore. It can be activated by using the --globals-backup option or setting backupGlobals="true" in the XML configuration file.

By using the --static-backup option or setting backupStaticAttributes="true" in the XML configuration file, this isolation can be extended to static attributes of classes.
```

**Comment by dkarlovi** [comment]
Good find, thanks, it should pass now, I'll merge as-is if you don't mind (can look over the changes if you have the time).

File: Tests/Security/Authentication/Provider/OAuthProviderTest.php [hunk 6]
```diff
@@ -35,7 +35,7 @@ class OAuthProviderTest extends \PHPUnit\Framework\TestCase
     protected $userProvider;
 
     /**
-     * @var \PHPUnit_Framework_MockObject_MockObject|OAuthProvider
+     * @var OAuthProvider
```
Comment (dinamic) [inline Tests/Security/Authentication/Provider/OAuthProviderTest.php]: Good catch.

**Comment by dinamic** [comment]
Changes in this PR make sense to me. Thanks for catching up 👍

File: Tests/Security/Authentication/Provider/OAuthProviderTest.php (hunk 6 above)
Comment (dkarlovi) [reply Tests/Security/Authentication/Provider/OAuthProviderTest.php]: I'd like to take credit, but it was PHPStan. :laughing:

File: composer.json [hunk 7]
```diff
@@ -17,26 +17,29 @@
     ],
     "require": {
         "php": "^7.1",
+        "alcaeus/mongo-php-adapter": "^1.1",
```
Comment (dkarlovi) [inline composer.json]: This 100% didn't show up in the PR before I merged it. :see_no_evil:

Comment (dinamic) [reply composer.json]: Maybe it got there while you were testing my proposed fix for phpunit v6.

Comment (dkarlovi) [reply composer.json]: I've hot-fixed it on master.

Comment (dinamic) [reply composer.json]: Don't you actually need that in order to use ODM now? We require PHP 7.1 which doesn't support the dated mongo extension anyway.

Comment (dkarlovi) [reply composer.json]: Yes, but we're only suggesting ODM bundle, not requiring it. It's up to the end-user to install the ODM bundle and, by which, this adapter.

Comment (dinamic) [reply composer.json]: Yea, you are right. Even if the end-user requires that workaround to use ODM, this is not the correct bundle to take care of the problem.

Comment (dkarlovi) [reply composer.json]: If you're not using Mongo ODM, no reason to have this installed.

Comment (stof) [reply composer.json]: yeah, this should be moved to the `require-dev` section

File: composer.json [hunk 8]
```diff
@@ -17,26 +17,29 @@
...
+        "phpstan/phpstan-shim": "~0.9",
+        "phpunit/phpunit": "~5.0|~6.0",
+        "propel/propel1": "~1.6",
+        "symfony/class-loader": "~3.0|~4.0",
+        "symfony/console": "~3.0|~4.0",
+        "symfony/form": "~3.0|~4.0",
+        "symfony/phpunit-bridge": "~3.0|~4.0",
+        "symfony/templating": "~3.0|~4.0",
+        "symfony/yaml": "~3.0|~4.0",
+        "willdurand/propel-typehintable-behavior": "~1.0"
```
Comment (stof) [inline composer.json]: that change is wrong. You are changing the lowest bound of the requirement

Comment (dkarlovi) [reply composer.json]: Thanks @stof, nice catch, will fix it.

**Pull request merged**